src/mito2/src/schedule/remote_job_scheduler.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Instant;
+
+use common_telemetry::error;
+use store_api::storage::RegionId;
+use tokio::sync::mpsc::Sender;
+
+use crate::compaction::compactor::CompactionRegion;
+use crate::compaction::picker::PickerOutput;
+use crate::error::Result;
+use crate::manifest::action::RegionEdit;
+use crate::request::{
+    BackgroundNotify, CompactionFailed, CompactionFinished, OutputTx, WorkerRequest,
+};
+
+pub type RemoteJobSchedulerRef = Arc<dyn RemoteJobScheduler>;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// RemoteJobScheduler is a trait that defines the API to schedule remote jobs.
+/// For example, a compaction job can be scheduled remotely as the following workflow:
+/// ```mermaid
+///   participant User
+///   participant MitoEngine
+///   participant CompactionScheduler
+///   participant Plugins
+///   participant RemoteJobScheduler
+///
+///   User->>MitoEngine: Initiates compaction
+///   MitoEngine->>CompactionScheduler: schedule_compaction()
+///   CompactionScheduler->>Plugins: Handle plugins
+///   CompactionScheduler->>RemoteJobScheduler: schedule(CompactionJob)
+///   RemoteJobScheduler-->>CompactionScheduler: Returns Job UUID
+///   CompactionScheduler-->>MitoEngine: Task scheduled with Job UUID
+///   MitoEngine-->>User: Compaction task scheduled
+/// ```
+#[async_trait::async_trait]
+pub trait RemoteJobScheduler: Send + Sync + 'static {
+    /// Sends a job to the scheduler and returns a UUID for the job.
+    async fn schedule(&self, job: RemoteJob, notifier: Arc<dyn Notifier>) -> Result<String>;
```

Comment:
Do we need to use `Arc` instead of `Box` for the notifier? Since `notify()` requires `&mut self` but you are hard `&mut self` from `Arc`.

I'd recommend changing it to `Box<dyn Notifier>`.

Reply:
TIL, you're right.